### ISSUE TITLE
"not" operator not working as expected with pe.version_info

### ISSUE THREAD

**Issue by tlansec** [VirusTotal/yara#1180]
Hello,

I was recently looking at the following file:

67757D0A55960716CF713557256BB21086774B513941FE9586EFC7139D42A336

This is a file with 1 signature and no version information. Here's some snipped output from the -D flag with yara:

```
signatures
		[0]
			thumbprint = "564e01066387f26c912010d06bd78d3cf1e845ab"
			issuer = "/C=US/ST=Washington/L=Redmond/O=Microsoft Corporation/OU=Copyright (c) 2000 Microsoft Corp./CN=Microsoft Code Signing PCA"
			subject = "/C=US/ST=Washington/L=Redmond/O=Microsoft Corporation/CN=Microsoft Corporation"
			version = 3
			algorithm = "sha1WithRSAEncryption"
			serial = "61:46:9e:cb:00:04:00:00:00:65"
			not_before = 1144179826
			not_after = 1191527626
[...snipped]
version_info
[...snipped (there is no version info, which is correct)
```

When I try to match on this using a rule like this, no match is provided - presumably because it cant find any version information:

```
import "pe" 
rule general_win_appended_microsoft_certificate
{
    condition:
        pe.signatures[0].subject contains "L=Redmond/O=Microsoft Corporation" and
        not pe.version_info["LegalCopyright"] contains "Microsoft"
}
```

I think in cases like this the signature should fire, thoughts?

Cheers,
Tom

**Comment by tlansec** [comment]
Alternatively, is there a way to write a rule for files which have no version information?

Cheers,
Tom

**Comment by plusvic** [comment]
This is interesting. The reason is that `pe.version_info["LegalCopyright"]` is actually `undefined`, because of this, the operation `pe.version_info["LegalCopyright"] contains "Microsoft"` is also `undefined`. The result from`not undefined` is also `undefined`. And finally the `and` between a `true` and `undefined` is false, because we can't tell for sure the second operand is true.

So, this is working as intended, but I agree that is confusing and maybe shouldn't work like that. Probably `undefined contains "Microsoft"` should be false.

Regarding the second question, there's no way for checking if a dictionary contains a key or not.

**Comment by tlansec** [comment]
Thanks for taking a look. I think either solution is acceptable (i.e. the above rule should work on the example file, or alternatively it should be possible to check for the existence of a key in a dictionary somehow).

Perhaps the second solution might have more benefits across other parts of YARA that have dictionary objects.

Cheers,
Tom

**Comment by plusvic** [comment]
Fixed in b7da5d2835cc4cf8a15027da30265addab1a4be5
